pkg/object/validator/validate.go

```diff
@@ -0,0 +1,186 @@
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/object/dependson"
+	"sigs.k8s.io/cli-utils/pkg/object/mutation"
+)
+
+// Validator contains functionality for validating a set of resources prior
+// to being used by the Apply functionality. This imposes some constraint not
+// always required, such as namespaced resources must have the namespace set.
+type Validator struct {
+	Mapper meta.RESTMapper
+}
+
+// Validate validates the provided resources. A RESTMapper will be used
+// to fetch type information from the live cluster.
+func (v *Validator) Validate(objs []*unstructured.Unstructured) *MultiValidationError {
+	crds := findCRDs(objs)
+	var allErrors []*ValidationError
+
+	ids := UnstructuredSetToObjMetadataSet(objs)
+
+	for i, obj := range objs {
+		id := ids[i]
+		var objErrors []error
+		if err := v.validateName(obj); err != nil {
+			objErrors = append(objErrors, err)
+		}
+		if err := v.validateNamespace(obj, crds); err != nil {
+			objErrors = append(objErrors, err)
+		}
+		if errs := v.validateDependsOnAnnotation(obj, ids); len(errs) > 0 {
+			objErrors = append(objErrors, errs...)
+		}
+		if errs := v.validateMutationAnnotation(obj, ids); len(errs) > 0 {
+			objErrors = append(objErrors, errs...)
+		}
+		if len(objErrors) > 0 {
+			allErrors = append(allErrors, &ValidationError{
+				Identifier: id,
+				Errors:     objErrors,
+			})
+		}
+	}
+	if len(allErrors) > 0 {
+		return &MultiValidationError{
+			Errors: allErrors,
+		}
+	}
+	return nil
+}
+
+// findCRDs looks through the provided resources and returns a slice with
+// the resources that are CRDs.
+func findCRDs(us []*unstructured.Unstructured) []*unstructured.Unstructured {
+	var crds []*unstructured.Unstructured
+	for _, u := range us {
+		if object.IsCRD(u) {
+			crds = append(crds, u)
+		}
+	}
+	return crds
+}
+
+// validateName validates the value of the name field of the resource.
+func (v *Validator) validateName(u *unstructured.Unstructured) error {
+	if u.GetName() == "" {
+		return field.Required(field.NewPath("metadata", "name"), "name is required")
+	}
+	return nil
+}
+
+// validateNamespace validates the value of the namespace field of the resource.
+func (v *Validator) validateNamespace(u *unstructured.Unstructured, crds []*unstructured.Unstructured) error {
+	scope, err := object.LookupResourceScope(u, crds, v.Mapper)
+	if err != nil {
+		return err
+	}
+
+	ns := u.GetNamespace()
+	if scope == meta.RESTScopeNamespace && ns == "" {
+		return field.Required(field.NewPath("metadata", "namespace"), "namespace is required")
+	}
+	if scope == meta.RESTScopeRoot && ns != "" {
+		return field.Invalid(field.NewPath("metadata", "namespace"), ns, "namespace must be empty")
+	}
+	return nil
+}
+
+func (v *Validator) validateDependsOnAnnotation(obj *unstructured.Unstructured, ids object.ObjMetadataSet) []error {
```

Comment:
Should we validate [CyclicDependencyError](https://github.com/kubernetes-sigs/cli-utils/blob/45a0b5b7e6b292c0a59c899aa50a3662a0c1cf7d/pkg/object/graph/graph.go#L139) here too?

Reply:
In order to find a cyclic error, we'd have to build the graph first. It seems pretty inefficient to do that twice. I'm not sure if we want to move graph building forward in front of validation... and building the graph might require valid inputs...

I guess we could make the graph SortObj emit events, but it doesn't do that now so it would require some significant changes...

Reply:
Alright, I've moved all the dependency and sorting errors to SortObjs and moved that forward in the applier so we can use the errors as validation errors, and then modified SortObjs to return a MultiError so we get more than one at a time.